### PR TITLE
[Refactor](1/N)get tablet schema from specific tablet metadata

### PR DESCRIPTION
**Pull request by sduzh**
This PR is part of the PR stack to support schema evolution without rewriting data, which require specifying the tablet version when fetching the tablet schema.

This PR mainly replaces the Tablet::get_schema() call in CompactionPolicy with TabletMetadataPB::schema().

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by sduzh** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 100 / 103 (97.09%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/compaction_policy.cpp|54|57|94.74%|[190, 270, 271]|
|:large_blue_circle:|src/storage/lake/vertical_compaction_task.cpp|9|9|100.00%|[]|
|:large_blue_circle:|src/storage/lake/horizontal_compaction_task.h|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/compaction_policy.h|5|5|100.00%|[]|
|:large_blue_circle:|src/storage/lake/rowset.cpp|10|10|100.00%|[]|
|:large_blue_circle:|src/storage/lake/vertical_compaction_task.h|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/compaction_task.cpp|2|2|100.00%|[]|
|:large_blue_circle:|src/service/service_be/lake_service.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|6|6|100.00%|[]|
|:large_blue_circle:|src/storage/lake/rowset_update_state.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/horizontal_compaction_task.cpp|7|7|100.00%|[]|
|:large_blue_circle:|src/storage/rowset/segment_rewriter.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/tablet.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/txn_log_applier.cpp|1|1|100.00%|[]|

**Comment by sduzh** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 100 / 103 (97.09%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/compaction_policy.cpp|54|57|94.74%|[190, 270, 271]|
|:large_blue_circle:|src/storage/lake/vertical_compaction_task.cpp|9|9|100.00%|[]|
|:large_blue_circle:|src/storage/lake/horizontal_compaction_task.h|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/compaction_policy.h|5|5|100.00%|[]|
|:large_blue_circle:|src/storage/lake/rowset.cpp|10|10|100.00%|[]|
|:large_blue_circle:|src/storage/lake/vertical_compaction_task.h|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/compaction_task.cpp|2|2|100.00%|[]|
|:large_blue_circle:|src/service/service_be/lake_service.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|6|6|100.00%|[]|
|:large_blue_circle:|src/storage/lake/rowset_update_state.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/horizontal_compaction_task.cpp|7|7|100.00%|[]|
|:large_blue_circle:|src/storage/rowset/segment_rewriter.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/tablet.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/txn_log_applier.cpp|1|1|100.00%|[]|

File: be/src/storage/lake/rowset.cpp [hunk 1]
```diff
@@ -27,10 +27,10 @@
 
 namespace starrocks::lake {
 
-Rowset::Rowset(Tablet* tablet, RowsetMetadataPtr rowset_metadata, int index)
+Rowset::Rowset(Tablet tablet, RowsetMetadataPtr rowset_metadata, int index)
         : _tablet(tablet), _rowset_metadata(std::move(rowset_metadata)), _index(index) {}
 
-Rowset::Rowset(Tablet* tablet, RowsetMetadataPtr rowset_metadata)
+Rowset::Rowset(Tablet tablet, RowsetMetadataPtr rowset_metadata)
```
[the remark this answers is not in the document]
Comment (sduzh) [reply be/src/storage/lake/rowset.cpp:33]: updated

File: be/src/storage/lake/rowset.h [hunk 2]
```diff
@@ -80,7 +81,7 @@ class Rowset {
 
 private:
     // _tablet is owned by TabletReader
-    Tablet* _tablet;
+    Tablet _tablet;
```
[the remark this answers is not in the document]
Comment (sduzh) [reply be/src/storage/lake/rowset.h:83]: updated

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 102 / 105 (97.14%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/compaction_policy.cpp|54|57|94.74%|[190, 270, 271]|
|:large_blue_circle:|src/storage/lake/vertical_compaction_task.cpp|9|9|100.00%|[]|
|:large_blue_circle:|src/storage/lake/horizontal_compaction_task.h|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/compaction_policy.h|5|5|100.00%|[]|
|:large_blue_circle:|src/storage/lake/rowset.cpp|12|12|100.00%|[]|
|:large_blue_circle:|src/storage/lake/vertical_compaction_task.h|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/compaction_task.cpp|2|2|100.00%|[]|
|:large_blue_circle:|src/service/service_be/lake_service.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|6|6|100.00%|[]|
|:large_blue_circle:|src/storage/lake/rowset_update_state.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/horizontal_compaction_task.cpp|7|7|100.00%|[]|
|:large_blue_circle:|src/storage/rowset/segment_rewriter.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/tablet.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/storage/lake/txn_log_applier.cpp|1|1|100.00%|[]|

File: be/src/storage/lake/compaction_policy.cpp [hunk 3]
```diff
@@ -532,25 +540,24 @@ StatusOr<CompactionAlgorithm> CompactionPolicy::choose_compaction_algorithm(cons
...
+        return std::make_shared<SizeTieredCompactionPolicy>(tablet_mgr, std::move(tablet_metadata));
+    } else {
+        return std::make_shared<BaseAndCumulativeCompactionPolicy>(tablet_mgr, std::move(tablet_metadata));
     }
-    return std::make_shared<BaseAndCumulativeCompactionPolicy>(std::move(tablet));
 }
 
-double compaction_score(const TabletMetadataPB& metadata) {
-    if (is_primary_key(metadata)) {
+double compaction_score(const std::shared_ptr<const TabletMetadataPB>& metadata) {
```
Comment (kevincai) [inline be/src/storage/lake/compaction_policy.cpp:559]: is it still necessary to pass shared_ptr by const reference? or pass by value is good enough?

**Review by kevincai — APPROVED** [review]

**Review by wyb — APPROVED** [review]

File: be/src/storage/lake/compaction_policy.cpp (hunk 3 above)
Comment (sduzh) [reply be/src/storage/lake/compaction_policy.cpp:559]: In theory, passing const reference here should still be better than passing by value, as it avoids the extra shared_ptr reference counting increment. Of course, this optimization may be negligible. However, passing const reference here should always be non-negative.

**Review by gengjun-git — APPROVED** [review]

**Pull request merged**